gateleen-routing/src/main/java/org/swisspush/gateleen/routing/Forwarder.java

```diff
@@ -176,6 +184,38 @@ private String extractUserId(HttpServerRequest request, Logger log) {
         }
     }
 
+    /**
+     * Execute the HeaderFunctions chain and apply the configured rules headers therefore
+     * @param log The logger to be used
+     * @param headers The headers which must be updated according the current forwarders rule
+     * @return null if everything is properly done or a error message if something went wrong.
+     */
+    String applyHeaderFunctions(final Logger log, MultiMap headers) {
+        final String hostHeaderBefore = HttpHeaderUtil.getHeaderValue(headers, HOST_HEADER);
+        final HeaderFunctions.EvalScope evalScope = rule.getHeaderFunction().apply(headers);
+        final String hostHeaderAfter = HttpHeaderUtil.getHeaderValue(headers, HOST_HEADER);
+        // see https://github.com/swisspush/gateleen/issues/394
+        if (hostHeaderAfter == null || hostHeaderAfter.equals(hostHeaderBefore)) {
+            // there was no host header before or the host header was not updated by the rule given,
+            // therefore it will be forced overwritten independent of the incoming value if necessary.
+            final String newHost = target.split("/")[0];
```

Comment:
Please note that `String#split()` is an expensive operation. Here you potentially extract the same hostname over and over again for every request.
Additionally I presume that the instance variable `target` only and always contains `host:port` - i.e. never something like an URL or URI. So you will _never_ find a forward `/` (though I'm not 100% sure). Therefore you probably can get rid of `String#split()`

Reply:
Hi oli :) good catch. I think alternatively this operation could also be done in the constructor. Then we could hold the "newHost" in a field and re-use the same string instance each time. I think this could help too.

Reply:
Agree. Note that this was already before executed each time. It shouldn't be worse than before therefore.

Reply:
Yes, seems that "port" can be dynamic (i.e. potentially change for each request) - still no need to search for a forward-slash as there is none ;-)

But thinking on this "dynamic/wildcard port"-stuff: Is currrent implementation really robust here? Note that one Forwarder-instance is create for one routing-rule (or Hook) but then used for multiple request. Be careful with an argument that Vert.x is single-threaded: yes it is, still different methods on one Forwarder-instance are called at different times during a request/response-cycle.
```int port``` should probably be _not_ an instance variable